bpf/lib/conntrack.h

```diff
@@ -62,17 +62,54 @@ enum {
 	ACTION_CLOSE,
 };
 
+union tcp_flags {
+#if defined(__LITTLE_ENDIAN_BITFIELD)
+	__u16	res1:4, doff:4, fin:1, syn:1, rst:1, psh:1, ack:1, urg:1, ece:1, cwr:1;
+#elif defined(__BIG_ENDIAN_BITFIELD)
+	__u16	doff:4, res1:4, cwr:1, ece:1, urg:1, ack:1, psh:1, rst:1, syn:1, fin:1;
+#else
+#error	"Adjust your <asm/byteorder.h> defines"
+#endif
+	struct {
+		__u8 upper_bits;
+		__u8 lower_bits;
+	};
+};
+
 static inline void __inline__ __ct_update_timeout(struct ct_entry *entry,
-						  __u32 lifetime)
+						  __u32 lifetime, int dir,
+						  union tcp_flags flags)
 {
+	__u32 now = bpf_ktime_get_sec();
+	__u8 *accumulated_flags;
+	__u8 seen_flags = flags.lower_bits;
+
 #ifdef NEEDS_TIMEOUT
-	entry->lifetime = bpf_ktime_get_sec() + lifetime;
+	entry->lifetime = now + lifetime;
 #endif
+	if (dir == CT_INGRESS) {
+		accumulated_flags = &entry->rx_flags_seen;
+	} else {
+		accumulated_flags = &entry->tx_flags_seen;
+	}
+	seen_flags |= *accumulated_flags;
+	if (*accumulated_flags != seen_flags) {
```

Comment:
We might want to exclude certain flags such as ECN, URG and PSH. Not a blocker but might help prevent unnecessary notifications.